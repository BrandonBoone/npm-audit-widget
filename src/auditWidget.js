const _render = (severities, title) => {
  const severityKeys = ['critical', 'high', 'moderate', 'low'];
  let numOfIssues = 'N/A';
  let footerText = 'no audit_results';
  let backgroundColor = ''
  let color = '#000000';

  if (severities) {
    numOfIssues = severityKeys.reduce((prev, next) => prev + (severities[next] || 0), 0);
    footerText = severityKeys.reduce((prev, next) => `${prev ? `${prev}, `: prev } ${next.charAt(0)}: ${(severities[next] || 0)}`, '');
    
    if (severities['critical']) {
      backgroundColor = '#DA0A00';
      color = '#ffffff';
    } else if (severities['high']) {
      backgroundColor = '#FF6201';
      color = '#ffffff';
    } else if (severities['moderate']) {
      backgroundColor = '#F08700';
      color = '#ffffff';
    } else if (severities['low']) {
      backgroundColor = '#F8A800';
      color = '#000';
    } else {
      backgroundColor = '#107C10';
      color = '#ffffff'
    }
  }
  return `
<div id='widget' class='widget' style='${color ? `color:${color};`: ''}${backgroundColor ? `background-color:${backgroundColor};`: ''}'>
  <h2 class="title">${title}</h2>
  <div class="big-count truncated-text-ellipsis">${numOfIssues}</div>
  <div class="footer truncated-text-ellipsis">${footerText}</div>
  <div class="footer truncated-text-ellipsis">
    <a style='${color ? `color:${color};`: ''}' href="https://docs.npmjs.com/getting-started/running-a-security-audit">npm audit</a> results
  </div>
</div>`;
}

const _getZipData = ({ client, buildId, projectId, JSZip }) =>
  client.getArtifactContentZip(buildId, 'audit_results', projectId)
  .then((zipArrayBuffer) => {
    if (!zipArrayBuffer) {
      return null;
    }
    return JSZip.loadAsync(zipArrayBuffer)
    .then((zip) => zip.file('audit_results/audit.json').async('string'))
    .then((auditJson) =>  JSON.parse(auditJson));
  });

const _showWarnings = ({ callback, VSS, JSZip, WidgetHelpers, BuildRestClient, BuildContracts }) =>
  (widgetSettings) => {
    var customSettings = JSON.parse(widgetSettings.customSettings.data);
    if (!customSettings) {
      return WidgetHelpers.WidgetStatusHelper.Success();
    }

    const projectId = VSS.getWebContext().project.id;       
    const client = BuildRestClient.getClient();

    client.getBuilds(projectId, [customSettings.definitionId])
    .then((builds) => {
      if (builds.length > 0) {
        return _getZipData({ client, buildId: builds[0].id, projectId, JSZip })
        .then((auditData) => callback(
          _render(auditData && auditData.metadata && auditData.metadata.vulnerabilities, customSettings.buildName),
          auditData
        ));
      }
    }); //todo: error handling
    return WidgetHelpers.WidgetStatusHelper.Success();
  }

export default (VSS, JSZip, callback) => {
  VSS.init({                        
    explicitNotifyLoaded: true,
    usePlatformStyles: true
  });

  VSS.require(
    ['TFS/Dashboards/WidgetHelpers', 'TFS/Build/RestClient', 'TFS/Build/Contracts'],
    (WidgetHelpers, BuildRestClient, BuildContracts) => {
      WidgetHelpers.IncludeWidgetStyles();
      const showWarnings = _showWarnings({ callback, VSS, JSZip, WidgetHelpers, BuildRestClient, BuildContracts });
      
      VSS.register('npmAuditWidget', () => ({  
        load: (widgetSettings) => showWarnings(widgetSettings),
        reload: (widgetSettings) => showWarnings(widgetSettings)
      }));
      VSS.notifyLoadSucceeded();
    }
  );
};
