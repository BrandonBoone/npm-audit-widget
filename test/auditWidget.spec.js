import auditWidget from '../src/auditWidget';
import MockVss from './VSS.mock';
import JSZip from 'jszip';
const path = require('path');

describe('auditWidget', () => {
  it('tid-0: handles no results', () => {
    expect.assertions(2);
    const VSS = new MockVss({ zipFile: null });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp, data) => {
        expect(data).toEqual(null);
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-1: handles moderate render', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: path.join('tid-1', 'audit_results.zip') });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-2: handles critical render', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: path.join('tid-2', 'audit_results.zip')  });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-3: handles high render', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: path.join('tid-3', 'audit_results.zip')  });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-4: handles low render', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: path.join('tid-4', 'audit_results.zip')  });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-5: handles ok render', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: path.join('tid-5', 'audit_results.zip')  });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-6: handles no builds', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: null, noBuilds: true });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-7: handles no settings', () => {
    expect.assertions(1);
    const VSS = new MockVss({ zipFile: null, noBuilds: true, noSettings: true });

    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        expect(renderedApp).toMatchSnapshot();
        resolve();
      })
    );
  });

  it('tid-8: handles reload', () => {
    expect.assertions(2);
    const VSS = new MockVss({ zipFile: path.join('tid-8', 'audit_results.zip')   });

    var called = 0;
    return new Promise((resolve) =>
      auditWidget(VSS, JSZip, (renderedApp) => {
        called++;
        if (called === 1) {
          expect(renderedApp).toMatchSnapshot();
          VSS.reload();
        } else if (called === 2) {
          expect(renderedApp).toMatchSnapshot();
          resolve();
        }
      })
    );
  });
});