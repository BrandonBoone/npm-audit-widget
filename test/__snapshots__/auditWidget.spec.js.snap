// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditWidget tid-0: handles no results 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#000000;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">N/A</div>
  <div class=\\"footer truncated-text-ellipsis\\">no audit_results</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#000000;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-1: handles moderate render 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#F08700;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">10</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 0,  m: 10,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-2: handles critical render 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#DA0A00;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">15</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 1,  h: 4,  m: 10,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-3: handles high render 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#FF6201;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">11</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 1,  m: 10,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-4: handles low render 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#000;background-color:#F8A800;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">1</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 0,  m: 0,  l: 1</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#000;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-5: handles ok render 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#107C10;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">0</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 0,  m: 0,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-6: handles no builds 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#000000;\\">
  <h2 class=\\"title\\">null</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">N/A</div>
  <div class=\\"footer truncated-text-ellipsis\\">no audit_results</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#000000;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-7: handles no settings 1`] = `
"<div id=\\"widget\\" class=\\"widget\\">
    <h2 class=\\"title\\">npm audit widget</h2>
    <div class=\\"big-count truncated-text-ellipsis\\">N/A</div>
    <div class=\\"footer truncated-text-ellipsis\\">not configured</div>
  </div>"
`;

exports[`auditWidget tid-8: handles reload 1`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#107C10;\\">
  <h2 class=\\"title\\">TRUNK</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">0</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 0,  m: 0,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;

exports[`auditWidget tid-8: handles reload 2`] = `
"
<div id=\\"widget\\" class=\\"widget\\" style=\\"color:#ffffff;background-color:#107C10;\\">
  <h2 class=\\"title\\">TRUNK2</h2>
  <div class=\\"big-count truncated-text-ellipsis\\">0</div>
  <div class=\\"footer truncated-text-ellipsis\\"> c: 0,  h: 0,  m: 0,  l: 0</div>
  <div class=\\"footer truncated-text-ellipsis\\">
    <a style=\\"color:#ffffff;\\" href=\\"https://docs.npmjs.com/getting-started/running-a-security-audit\\">npm audit</a> results
  </div>
</div>"
`;
